### ISSUE TITLE
holtWintersAberration no longer works correctly in 0.9.16

### ISSUE THREAD

**Issue by bemeyert** [graphite-project/graphite-web#1957]
Hi all,

We just upgraded from 0.9.15 to 0.9.16 and our graphs using holtWintersAberration no longer work correctly. When we set "from" to less than 9 days we get the following error:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python2.7/site-packages/graphite/render/views.py", line 123, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/usr/lib/python2.7/site-packages/graphite/render/evaluator.py", line 10, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/usr/lib/python2.7/site-packages/graphite/render/evaluator.py", line 21, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/usr/lib/python2.7/site-packages/graphite/render/evaluator.py", line 31, in evaluateTokens
    return func(requestContext, *args)
  File "/usr/lib/python2.7/site-packages/graphite/render/functions.py", line 2078, in holtWintersAberration
    elif upperBand[i] is not None and series[i] > upperBand[i]:
IndexError: list index out of range
```
The request looks like this:
```
http://metrics.server/render?&rawData=true&format=json&from=-2d&until=now&target=holtWintersAberration(groupByNode(scaleToSeconds(serieslist,%201),%207,%20%27sum%27),%2010)
```
This worked fine in 0.9.15. Since we use CentOS 7 the latest available version is 0.9.16. 

Thx for any help

**Comment by deniszh** [comment]
Looks like it's caused by https://github.com/graphite-project/graphite-web/commit/2e940a34d9aae0a27d1629dd1db8255763aeef51
which was backport of fix for issue #220 from a master.
So, I think version 1.0.x also contain this. 
Not sure how that patch could cause that, though. `series[i]` is defined, otherwise it should fail on line 1975 instead, so it's caused by bad `upperBand[i]` - but it was there before and reverting this patch doesn't help.
And I do not see any other changes in holtWintersAberration.
So, I'm suspect that you just getting NULLs in your data when getting older data, maybe because of wrong aggregation policy. Could you please check that? just remove holtWintersAberration and check rest of your data.

**Comment by bemeyert** [comment]
@deniszh All other queries using that series work fine with `from` set to less than 9 days. E.g.
```
groupByNode(scaleToSeconds(series 1), 7, 'sum')
holtWintersForecast(groupByNode(scaleToSeconds(series, 1), 7, 'sum'))
holtWintersConfidenceBands(groupByNode(scaleToSeconds(series, 1), 7, 'sum'), 10)
```
Only Aberration is making problems.

**Comment by deniszh** [comment]
Hmm... Could you please check if reverting https://github.com/graphite-project/graphite-web/commit/2e940a34d9aae0a27d1629dd1db8255763aeef51 helps then?
If not - then it's some other change, probably in  `groupByNode()` then.

**Comment by bemeyert** [comment]
I tried that. First the diff:
```
# diff /usr/lib/python2.7/site-packages/graphite/render/functions.py /usr/lib/python2.7/site-packages/graphite/render/functions.py.orig 
2078,2079c2078
<       #elif upperBand[i] is not None and series[i] > upperBand[i]:
<       elif series[i] > upperBand[i] :
---
>       elif upperBand[i] is not None and series[i] > upperBand[i]:
2081,2082c2080
<       #elif lowerBand[i] is not None and series[i] < lowerBand[i]:
<       elif series[i] < lowerBand[i] :
---
>       elif lowerBand[i] is not None and series[i] < lowerBand[i]:
```
And the the request: `http://metrics/render?&rawData=true&format=json&from=-2d&until=now&target=holtWintersAberration(groupByNode(scaleToSeconds(series,%201),%207,%20%27sum%27),%2010)`
Except the error message nothing has changed.

**Comment by deniszh** [comment]
Thanks, looks like the problem lies somewhere else. I checked `scaleToSeconds()` and `groupByNode()` functions too - there was no changes there in 0.9.16 either.

**Comment by bemeyert** [comment]
When looking with `git difftool 0.9.15 0.9.16 webapp/graphite/render/functions.py` I can see a lot of changes in the `holtWintersConfidenceBands` function. Maybe the order or the structure of the returned data differs between the versions?

**Comment by deniszh** [comment]
Ah, you're right, completely missed that.
It's coming from https://github.com/graphite-project/graphite-web/pull/1523 by @arielnh56 
He mentioned that there's problem with 7 days selection, but 9 days should not be affected:
"holtWinters functions old version had a discontinuity around a 7 day
time selection. 167 hours significantly different from 169 hours. This
discontinuity is now gone, but the results for shorter time windows are
now significantly changed. I'm not a statistician but I suspect these
are intended for long time windows and the discontinuity was a bug
brought on by mixing two whisper ranges together."

**Comment by bemeyert** [comment]
Ahh, this explains the difference with 0.9.15. That leaves us currently with either partially wrong data (0.9.15) or a timerange/graph that is too big (0.9.16). I'll meditate on this for a while... ;)

**Comment by arielnh56** [comment]
I figured the discontinuity could not be right, but frankly I don't know
what holtWintersAberration output is supposed to be like.

What I changed was the method of re-retrieving the data, which is required
as the time period for the function starts prior to the output window. In
this case the pre-time is fixed at 7 days.

In our environment we have default

retentions = 10s:2d,1m:7d,10m:2y

So at 7 days (168 hours) the data retrieved switches from 1m to 10m buckets.

What I *think* was happening is that in the old method
(_fetchWithBootstrap) for times under 7d is would pull an additional 7d
from the 1m bucket and glop that on the beginning of the 1m data, resulting
in a data series that was based on mixed bucket sizes. fetchWithBootstrap
did try to compensate for this by mapping the bootsrap data in some way
that probably worked for things like movingAverage but gave a radically
different result for holtWintersAberration

So in the case where the preview period extends across the retention
boundary, the new method uses the data from the retention corresponding to
the older low-res bucket, while the old one tried to map the data from the
low-res bucket onto the hi-res bucket granularity.

there was this:

if bootstrap.step != original.step:
      ratio = bootstrap.step / original.step
      for value in bootstrap:
         #XXX For series with aggregationMethod = sum this should also
         # divide by the ratio to bring counts to the same time unit
         # ...but we have no way of knowing whether that's the case
         newValues.extend([ value ] * ratio)
     else:
       newValues.extend(bootstrap)

Anyhoo when I was going through the results of my underlying change I
noticed this change in the hWA results and when I dug in I saw this radical
change in the *old* output when I changed the time from -167h to -169h. One
should see an offset graph, not a different graph.

So if you are wanting holtWintersAberration at any given granularity, I
think if you increase your retention for the buckets to at least 7 days
longer than you want to view, you should get the results that are what you
need. holtWintersAberration is, apparently, sensitive to bucket size (or
aggregation method?)

Probably the discrepancy I saw was because I was looking at random series
data which was set up with the default (average) aggregation method. There
may be a better aggregation method for data that wants to be looked at with
hWA.

**Comment by arielnh56** [comment]
edit

What I *think* was happening is that in the old method
(_fetchWithBootstrap) for times under 7d is would pull an additional 7d
from the *10m* bucket and glop that on the beginning of the 1m data

**Comment by bemeyert** [comment]
@arielnh56 Thanks for your answer. But I'm still a little bit confused. What is the behavior of your new code? We have configured a retention of `10s:28h, 1m:8d, 10m:750d`. As I wrote earlier I can get valid data only with -9 days. What I'd like is to view the last 6 or 12 hours.

**Comment by arielnh56** [comment]
The difference between the old code and the new code is how the
pre-time-period data is retrieved. Some functions need data before the
requested period in order to calculate the first datapoints. e.g. if you
have a daily moving average, you need the data for one day prior to time
period in order to calculate the

The old code used a routine called _fetchWithBootstrap to retrieve the
pre-time data. The problem I was fixing was that in our federated cluster
this occurs after the wildcard requests to the storage nodes are expanded.
While the original data request could have been a single call to the back
end, this "go back for more" was generating hundreds of calls on large
aggregate graphs leading to performance issues. Sometimes the graphs would
not render at all. My need came from movingAverage but I saw the same
inefficient code was being used by the holt-winters funtions.

The new code passes the wildcard meta-data to to the function and it uses
the wildcard to retrieve the data once more with the earlier start point.
If the start point is now in a lower resolution time buck, all the data
retrieved will be from the lower resolution bucket.

The holt-winters problem arises as follows:


   1. It has a fixed pre-fetch of 7 days.
   2. For recent-period graphs, this forces the pre-fetch to be pulled from
   a non-recent bucket
   3. _fetchWithBootstrap tries to map the older data to the new by
   reversing the aggregation process, assuming that sum() was the aggregation.
   This is clunky and at best inaccurate.
   4. This generated the discontinuity in behavior I observed at the bucket
   boundaries.


I think the fixed 7 day pre-fetch on the HWA is the outstanding problem
here. It implies that holt-winters needs 7 days prior data to build any
graph. I am not a statistician, and google doesn't immediately show up such
a link, so I'm thinking that that was an assumption/convenience for the
author of these functions that does not apply to every case. How much prior
data does you 6-12 hour graph need? Ideally the function should set that as
the pre-fetch and the bucket period should extend to include the ask time
and the pre-fetch. Your bucket period is 28h so that's fine. It's the 7
days that is forcing you into the low-res bucket

I transcribed the 7-day prefetch to the new calls in

def holtWintersForecast(requestContext, seriesList):
previewSeconds = 7 * 86400 # 7 days

def holtWintersConfidenceBands(requestContext, seriesList, delta=3):
previewSeconds = 7 * 86400 # 7 days




I *think* these values should not be constants. They should be calculated
to match the request period.

As a quick hack for your problem, I would try dropping previewSeconds to 15
hours (54000), this should allow your 12 hour requests with a 15 hour
pre-fetch all to come from your 28h:10s bucket.

**Comment by deniszh** [comment]
If `previewSeconds` is a problem - we can make it configurable, it's easy and backward compatible.
A key component of Holt-Winters method is seasonality, so, I think 7 days was chosen as a minimal period which should have good seasonality.

**Comment by deniszh** [comment]
PS: Very nice explanation of what Holt-Winters  is in [Grisha's blog](https://grisha.org) - [Part I](https://grisha.org/blog/2016/01/29/triple-exponential-smoothing-forecasting/), [Part II](https://grisha.org/blog/2016/02/16/triple-exponential-smoothing-forecasting-part-ii/), [Part III](https://grisha.org/blog/2016/02/17/triple-exponential-smoothing-forecasting-part-iii/)

**Comment by bemeyert** [comment]
@arielnh56 Thanks for your very good explanation. I'm slowly grasping what's happening here.
@deniszh It might be a good idea to make `previewSeconds` variable. If it should be configurable I can't tell, since my statistic-foo is not good enough. I first want to dig through the article you posted before I decide where I go from here. I'll post here what came out of it ;)
I want to thank you all for your time and effort this far.

Peace ;)

**Comment by deniszh** [comment]
@bemeyert : Let me know if changing `previewSeconds` helps in your particular case.
I'll submit PR for making it configurable.

**Comment by bemeyert** [comment]
@deniszh I'll get back to you as soon as I can. We have a math guy in our company and I try to get him involved ;)

**Comment by bemeyert** [comment]
@deniszh It would be great if we could set `previewSeconds` dynamically. We then could play around with ConfidenceBands and see what works best for our case. We want to see if there are drastic changes in a metric in the last minute.

**Comment by bemeyert** [comment]
Just one thing is still bugging me. I can use ConfidenceBands in any timeframe (-6h, -2d ...) and it works. But I get the error when I use Aberration. So, would it really fix our problem if `previewSeconds` is configurable?

**Comment by DanCech** [comment]
The best solution is likely to add `previewSeconds` as an additional optional parameter on the holt-winters functions.

**Comment by deniszh** [comment]
Yep, will do exactly that.

**Comment by deniszh** [comment]
Change should be as simple as https://github.com/graphite-project/graphite-web/pull/1961
Easy portable to 1.0.x and even 0.9.x.
Let me check it first, though.

**Comment by deniszh** [comment]
Or it's better to make it seconds instead of days? @bemeyert @DanCech (fractional days will work, of course, it only matters of convenience).

**Comment by bemeyert** [comment]
@deniszh I think that seconds as unit would be better. I.e. we have our first aggregation at 28 hours and I don't want to calculate the fractional days for that. It's much too error prone.

**Comment by deniszh** [comment]
@bemeyert : maybe hours is better then?

**Comment by bemeyert** [comment]
@deniszh hours or seconds. I guess both would work. Although I would prefer seconds.

**Comment by deniszh** [comment]
@bemeyert - does that looks better - https://github.com/graphite-project/graphite-web/pull/1965 ?
It will accept any Graphite compatible time string as bootstrap parameter. Default is '7d', you can use e.g. '15h'.

**Comment by bemeyert** [comment]
@deniszh Even better since one can choose. Thanks a lot! It should be easy to patch our current installation until a 1.x release is available for RHEL-based distributions.
I will try out the patch in the course of next week and update this issue.

**Comment by deniszh** [comment]
@bemeyert - thatnks! Let me know is it useful, I'll cherry-pick it to 1.0.x and 0.9.x branches.

**Comment by bemeyert** [comment]
@deniszh It works now. While looking at the last hour we use the query `holtWintersAberration(groupByNode(scaleToSeconds(metrics, 1), 7, 'sum'), 10, '26h')`. Works like a charm. If there are enough data for Holt-Winters only time will tell ;)

In which releases will this patch appear?

**Comment by deniszh** [comment]
@bemeyert Will cherry pick it now to 1.0.x branch, so, release 1.0.2, I think. Together with Django 1.8 support, (which should be in the same release), it should make porting it to RHEL/CentOS 7 quite easy.

**Comment by bemeyert** [comment]
So there will be no 0.9.x release anymore? Just asking. I'd like to use 1.0.x. So the sooner it hits EPEL the better ;)

**Comment by deniszh** [comment]
Yes, 0.9.16 was final. I can backport patch to 0.9.x branch, just for sake of completeness, but we're not going to release it.

**Comment by bemeyert** [comment]
The backport is not necessary. The next RPM will be a 1.0.x release. So long the manual patch stays on the system. A big thank you for your help and excellent work!

**Comment by deniszh** [comment]
Done - merged to master and 1.0.x branch, so, should be included in next release.
